File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        '''
        Time: O(n)
        Space: O(n)
        '''
        stack = []
        pairs = {
            ')': '(',
            '}': '{',
            ']': '['
        }
        for char in s:
            if char in "({[":
                stack.append(char)
            elif stack and stack[-1] == pairs[char]:
                stack.pop()
            else:
                return False
        return len(stack) == 0

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
private: 
    stack<char> st;
    unordered_map<char, char> close_to_open = {
        {')', '('},
        {']', '['},
        {'}', '{'}
    };

public:
    bool isValid(string s) {
        if (s.size() % 2 == 1)
            return false;
        
        for (auto ch : s) {
            if (ch == '(' || ch == '[' || ch == '{') {
                st.push(ch);
            }
            else if (st.size() == 0 || st.top() != close_to_open[ch]) {
                return false;
            }
            else {
                st.pop();
            }
        }

        return st.size() == 0;
    }
};